File: week-01-neuronal-models/mtlb/HodgkinHuxley.m
function dudt = HodgkinHuxley(t,u,p,IAppFun)
  % function dudt = HodgkinHuxley(t,u,p,IAppFun)
  % Computes the vectorfield of the Hodgkin-Huxley model
  % The notation follows ET, Chapter 1.10.
  % Input varaibles: 
  %      t: time, expressed as a scalar
  %      u: the vector variables u = [v; n; m; h]
  %      p: the control parameters p = [cm gNa gK gL eNa eK eL phi]
  %      IAppFun: a function handle that specifies the applied input current as a function of time
  % Output varaibles: 
  %      dudt: the vectorfield, with components ordered as in u

  %% Unpack parameters
  Cm   = p(1); 
  gNa  = p(2); 
  gK   = p(3); 
  gL   = p(4); 
  eNa  = p(5); 
  eK   = p(6); 
  eL   = p(7); 
  phi  = p(8); 

  %% Unpack state variables
  v = u(1);
  n = u(2);
  m = u(3);
  h = u(4);

  %% Ativation and inactivation functions
  alphaN = 0.01*(v+55)/(1-exp(-(v+55)/10));
  betaN  = 0.125*exp(-(v+65)/80);
  alphaM = 0.1*(v+40)/(1-exp(-(v+40)/10));
  betaM  = 4*exp(-(v+65)/18);
  alphaH = 0.07*exp(-(v+65)/20);
  betaH  = 1/(1+exp(-(v+35)/10));

  %% Compute currents
  INa  = -gNa*m^3*h*(v-eNa);
  IK   = -gK*n^4*(v-eK);
  IL   = -gL*(v-eL);
  IApp =  IAppFun(t);

  %% Assemble the vectorfield 
  dudt = zeros(4,1);
  dudt(1) = (INa + IK + IL + IApp)/Cm;
  dudt(2) = phi*(alphaN*(1-n) - betaN*n);
  dudt(3) = phi*(alphaM*(1-m) - betaM*m);
  dudt(4) = phi*(alphaH*(1-h) - betaH*h);

end

File: week-01-neuronal-models/py/HodgkinHuxley.py
from numpy import exp, zeros, array, ndarray

def HodgkinHuxley(
        t:float, # time
        # these are the (state determining) variables that are solved for
        variables:list[float],
        # these params determine the behavior of the model
        params:list[float],
        # function that determines the applied current I based on the timepoint (t)
        Ifunc:callable
        # returns du/dt with u containing state variables u = [v, n, m, h]
        ) -> ndarray:
    
    # v (voltage), n (activation gate Kv), m (activation gate Nav), h (inactivation gate Nav)
    v, n, m, h = variables
    # Cm (membrane capacitance), gNa (Na conductance), gK (K conductance), gL (leak conductance), 
    # eNa (Na Nernst), eK (K Nernst), eL (leak reversal potential), 
    # phi (temperature factor, influences transitions to different activation / inactivation states)
    Cm, gNa, gK, gL, eNa, eK, eL, phi = params

    # Calculate what the currents are from the state variables & parameters
    # IK = -gK * n^4(V-Ek)
    IK = - gK * (n**4) * (v - eK)
    # INa = -gNa * m^3 h(V-Ena)
    INa = - gNa * (m**3) * h * (v - eNa)
    # IL = -gL(V - El)
    IL = - gL * (v - eL)
    # Applied current at time t
    I_applied = Ifunc(t)

    # Activation & Inactivation functions  (parameters set by Hodkin & Huxley)
    # for the next timestep update in activation states 
    # ! based on current activation state (n /m /h) & voltage (v) !
    # in general these parameters are described by:
    # m = alpha_m(V)*(1-m) - beta_m(V)*m , same for n and h
    # Kv activation (n)
    alpha_n = 0.01 * (v + 55)/(1 - exp(-(v + 55)/10))
    beta_n  = 0.125 * exp(-(v + 65)/80)
    # Nav activation (m)
    alpha_m = 0.1 * (v + 40)/(1 - exp(-(v + 40)/10))
    beta_m  = 4 * exp(-(v + 65)/18)
    # Nav INactivation (h)
    alpha_h = 0.07 * exp(-(v + 65)/20)
    beta_h  = 1/(1 + exp(-(v + 35)/10))
    
    # du/dt
    # step in membrane voltage / step in time = Currents (t) / Membrane Capacitance
    dvdt = (IK + INa + IL + I_applied) / Cm
    # Changes in activation / inactivation states of channels -> this makes the difference, phi: temperature dependent!
    # Kv activation (n)
    dndt = phi * ((alpha_n * (1-n)) - (beta_n * n))
    # Nav activation (m)
    dmdt = phi * ((alpha_m * (1-m)) - (beta_m * m))
    # Nav INactivation (h)
    dhdt = phi * ((alpha_h * (1-h)) - (beta_h * h))

    # breakpoint()

    return array([dvdt,
                  dndt,
                  dmdt,
                  dhdt])